src/wallet/mod.rs

```diff
@@ -121,15 +112,6 @@ pub enum AddressIndex {
     /// Use with caution, if an index is given that is less than the current descriptor index
     /// then the returned address may have already been used.
     Peek(u32),
-    /// Return the address for a specific descriptor index and reset the current descriptor index
-    /// used by `AddressIndex::New` and `AddressIndex::LastUsed` to this value.
-    ///
-    /// Use with caution, if an index is given that is less than the current descriptor index
-    /// then the returned address and subsequent addresses returned by calls to `AddressIndex::New`
-    /// and `AddressIndex::LastUsed` may have already been used. Also if the index is reset to a
-    /// value earlier than the [`crate::blockchain::Blockchain`] stop_gap (default is 20) then a
-    /// larger stop_gap should be used to monitor for all possibly used addresses.
-    Reset(u32),
```

Comment:
The reason is to keep the invariant that every UTXO you have must have a derivation index less than or equal to your current derivation index. Clearly this cannot hold if you allow "resetting" your derivation index. Why do we want this invariant? It simplifies API and internals. There is no longer two indexes you are tracking: (i) where you are cached up to, (ii) where you are giving out addresses at. You just have one number which does handles both things.

However we should allow for "trimming" your derivation index down to your last active index. This will be needed in order to do `stop_gap` style syncing with things like CBF and `rpc` block streaming. The algorithm would be to set your derivation index to `last_active_index + stop_gap` before applying each block. Then when you've reached the tip you can trim to the `last_active_index` and the wallet becomes operational.

Reply:
Personally I think we should re-enable this feature to reset to the largest of:

1. First unused index
2. Parameter in `Reset(param)`

It shouldn't be complicated to implement with the `KeychainTxOutIndex` API.

Reply:
I think we would need to "unreveal" method in `KeychainTxOutIndex` first I think. I still find this very questionable. Why would you reveal an address and then not actually reveal it. In any case this feature shouldn't be done as an `AddressIndex`.